File: app/src/main/java/com/s95ammar/weeklyschedule/views/activities/MainActivity.java
package com.s95ammar.weeklyschedule.views.activities;

import android.content.Intent;
import androidx.annotation.NonNull;
import com.google.android.material.navigation.NavigationView;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.appcompat.app.ActionBarDrawerToggle;
import android.os.Bundle;
import androidx.appcompat.widget.Toolbar;
import android.text.format.DateFormat;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.s95ammar.weeklyschedule.models.Category;
import com.s95ammar.weeklyschedule.models.Schedule;
import com.s95ammar.weeklyschedule.models.SchedulesList;
import com.s95ammar.weeklyschedule.views.fragments.CategoriesListFragment;
import com.s95ammar.weeklyschedule.views.fragments.CategoryRefactorDialog;
import com.s95ammar.weeklyschedule.views.fragments.ScheduleNamerDialog;
import com.s95ammar.weeklyschedule.R;
import com.s95ammar.weeklyschedule.views.fragments.ScheduleViewerFragment;
import com.s95ammar.weeklyschedule.views.fragments.SchedulesListFragment;

import yuku.ambilwarna.AmbilWarnaDialog;


public class MainActivity extends AbstractScheduleActivity implements
        NavigationView.OnNavigationItemSelectedListener,
        SchedulesListFragment.SchedulesListManager,
        ScheduleNamerDialog.ScheduleNamerListener,
        CategoriesListFragment.CategoriesListManager,
        CategoryRefactorDialog.CategoryRefactor {

    private static final String TAG = "MainActivity";
    private DrawerLayout drawer;
    private SchedulesListFragment schedulesListFragment;
    private CategoriesListFragment categoriesListFragment;

    private interface NavDrawerItems {
        int ACTIVE_SCHEDULE = 0;
        int SCHEDULES = 1;
        int SETTINGS = 2;
        int INFO = 3;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        loadData();
        Toolbar toolbar = findViewById(R.id.toolbar_main_activity);
        setSupportActionBar(toolbar);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.toolbar_menu, menu);
        super.menu = menu;
        Toolbar toolbar = findViewById(R.id.toolbar_main_activity);
        checkActiveSchedule();
        setUpNavDrawer(toolbar);
        return true;
    }

    private void checkActiveSchedule() {
        if (SchedulesList.getInstance().hasActiveSchedule()) {
            Log.d(TAG, "checkActiveSchedule: Active schedule found " + SchedulesList.getInstance().getActiveSchedule());
            openScheduleViewerFragment(SchedulesList.getInstance().getActiveSchedule(), ScheduleViewerFragment.VIEW, R.id.fragment_container_main_activity);
        } else {
            Log.d(TAG, "checkActiveSchedule: Active schedule not found");
            switchToFragment(schedulesListFragment != null ? schedulesListFragment : (schedulesListFragment = new SchedulesListFragment()),
                    R.id.fragment_container_main_activity, null);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_APPLY_CHANGES) {
            if (resultCode == RESULT_OK) {
                applyEventChanges(data.getIntExtra(KEY_SCHEDULE_INDEX, -1), R.id.fragment_container_main_activity);
            }
        }

    }


    private void setUpNavDrawer(Toolbar toolbar) {
        drawer = findViewById(R.id.drawer_layout);
        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        if (scheduleViewerFragment != null && scheduleViewerFragment.getUserVisibleHint()) {
            navigationView.getMenu().getItem(NavDrawerItems.ACTIVE_SCHEDULE).setChecked(true);
        } else if (schedulesListFragment != null && schedulesListFragment.getUserVisibleHint()) {
            navigationView.getMenu().getItem(NavDrawerItems.SCHEDULES).setChecked(true);
        }
    }

    @Override
    public void onBackPressed() {
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        switch (menuItem.getItemId()) {
            case R.id.nav_active_schedule:
                openScheduleViewerFragment(SchedulesList.getInstance().getActiveSchedule(), ScheduleViewerFragment.VIEW, R.id.fragment_container_main_activity);
                break;
            case R.id.nav_schedules:
                switchToFragment(schedulesListFragment != null ? schedulesListFragment : (schedulesListFragment = new SchedulesListFragment()),
                        R.id.fragment_container_main_activity, null);
                break;
            case R.id.nav_categories:
                switchToFragment(categoriesListFragment != null ? categoriesListFragment : (categoriesListFragment = new CategoriesListFragment()),
                        R.id.fragment_container_main_activity, null);
                break;
            case R.id.nav_info:
                Intent intentInfo = new Intent(this, InfoActivity.class);
                startActivity(intentInfo);
                break;
        }
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    public void showScheduleInActivity(int i) {
        Intent intent = new Intent(this, ScheduleViewerActivity.class);
        intent.putExtra(SchedulesListFragment.SchedulesListManager.KEY_INDEX, i);
        startActivity(intent);
    }

    @Override
    public void applyName(String name, int i) {
        switch (i) {
            case ScheduleNamerDialog.Action.ADD:
                schedulesListFragment.addSchedule(name);
                break;
            default:
                schedulesListFragment.renameSchedule(name, i);
        }
    }

    @Override
    public void showScheduleRefactorDialog(String name, int i) {
        ScheduleNamerDialog dialog = new ScheduleNamerDialog();
        Bundle values = new Bundle();
        values.putInt(SchedulesListFragment.SchedulesListManager.KEY_INDEX, i);
        values.putString(SchedulesListFragment.SchedulesListManager.KEY_NAME, name);
        dialog.setArguments(values);
        dialog.show(getSupportFragmentManager(), TAG);
    }

    @Override
    public void showCategoryRefactorDialog(Category category, int i) {
        CategoryRefactorDialog dialog = new CategoryRefactorDialog();
        Bundle values = new Bundle();
        values.putInt(CategoriesListFragment.CategoriesListManager.KEY_INDEX, i);
        values.putSerializable(CategoriesListFragment.CategoriesListManager.KEY_CATEGORY, category);
        dialog.setArguments(values);
        dialog.show(getSupportFragmentManager(), TAG);
    }

    @Override
    public void applyCategory(Category category, int i) {
        switch (i) {
            case CategoryRefactorDialog.Action.ADD:
                categoriesListFragment.addCategory(category);
                break;
            default:
                categoriesListFragment.editCategory(category, i);
        }
    }

    @Override
    public void openColorPicker(final CategoryRefactorDialog categoryRefactorDialog, final int viewId, final int defaultColor) {
        AmbilWarnaDialog colorPicker = new AmbilWarnaDialog(this, defaultColor, new AmbilWarnaDialog.OnAmbilWarnaListener() {
            @Override
            public void onCancel(AmbilWarnaDialog dialog) {
            }

            @Override
            public void onOk(AmbilWarnaDialog dialog, int color) {
                Bundle bundle = new Bundle();
                bundle.putInt(KEY_ID, viewId);
                bundle.putInt(KEY_COLOR, color);
                categoryRefactorDialog.setArguments(bundle);
                categoryRefactorDialog.receiveColor();
            }
        });
        colorPicker.show();
    }

}

File: app/src/main/java/com/s95ammar/weeklyschedule/views/activities/MainActivity.kt
package com.s95ammar.weeklyschedule.views.activities

import android.os.Bundle
import android.view.MenuItem
import androidx.core.os.bundleOf
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavController
import androidx.navigation.NavDestination
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI.onNavDestinationSelected
import androidx.navigation.ui.navigateUp
import androidx.navigation.ui.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import com.google.android.material.navigation.NavigationView
import com.s95ammar.weeklyschedule.R
import com.s95ammar.weeklyschedule.models.data.Schedule
import com.s95ammar.weeklyschedule.util.close
import com.s95ammar.weeklyschedule.util.closeIfOpen
import com.s95ammar.weeklyschedule.util.isOpen
import dagger.android.support.DaggerAppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*
import javax.inject.Inject


class MainActivity : DaggerAppCompatActivity(), NavController.OnDestinationChangedListener, NavigationView.OnNavigationItemSelectedListener {

	@Inject
	lateinit var factory: ViewModelProvider.Factory
	private lateinit var navController: NavController
	private lateinit var appBarConfig: AppBarConfiguration
	private val topLevelDestinations = setOf(R.id.nav_top_level_schedule_viewer, R.id.nav_top_level_schedules, R.id.nav_top_level_categories)

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_main)
		initNavController()
	}

	private fun initNavController() {
		appBarConfig = AppBarConfiguration(topLevelDestinations, drawer_layout)
		navController = findNavController(R.id.nav_host_fragment)
		navController.setGraph(R.navigation.nav_graph, bundleOf(resources.getString(R.string.key_schedule_id) to Schedule.activeScheduleId))
		setupActionBarWithNavController(navController, appBarConfig)
		nav_view.setupWithNavController(navController)
		navController.addOnDestinationChangedListener(this)
		nav_view.setNavigationItemSelectedListener(this)
	}


	override fun onNavigationItemSelected(navigationItem: MenuItem): Boolean {
		return if (navController.currentDestination?.id != navigationItem.itemId) {
			if (topLevelDestinations.contains(navigationItem.itemId)) navController.popBackStack()
			when (navigationItem.itemId) {
				R.id.nav_top_level_schedule_viewer -> {
					navController.navigate(R.id.nav_top_level_schedule_viewer, bundleOf(resources.getString(R.string.key_schedule_id) to Schedule.activeScheduleId))
					drawer_layout.closeIfOpen()
					true
				}
				else -> onNavDestinationSelected(navigationItem, navController).also { handled ->
					if (handled) drawer_layout.closeIfOpen()
				}
			}
		} else {
			drawer_layout.closeIfOpen()
			false
		}
	}

	override fun onDestinationChanged(controller: NavController, destination: NavDestination, arguments: Bundle?) {
		drawer_layout.setDrawerLockMode(if (topLevelDestinations.contains(destination.id)) DrawerLayout.LOCK_MODE_UNLOCKED else DrawerLayout.LOCK_MODE_LOCKED_CLOSED)
	}


	// adds functionality to burger icon (only for opening the drawer) and back arrow
	override fun onSupportNavigateUp() = navController.navigateUp(appBarConfig) || super.onSupportNavigateUp()

	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		when (item.itemId) {
			android.R.id.home ->
				if (drawer_layout.isOpen()) {
					drawer_layout.close()
					return true
				}
		}
		return false
	}

	override fun onBackPressed() {
		if (drawer_layout.isOpen())
			drawer_layout.close()
		else
			super.onBackPressed()
	}

}
